File: old/wrapped.rb
# stats per person
# - total upvotes
# - total downvotes
# - net score
# - biggest critics (3) - people who gave you the most downvotes
# - biggest fans (3) - people who gave you the most upvotes
# - best songs (3) - songs with highest net score
# - worst songs (3) - songs with lowest net score
# - total characters typed
# - top artists submitted (3) - artist names you've submitted the most frequently

# overall stats
# - most upvoted songs (top 3)
# - most downvoted songs (top 3)
# - person with most upvotes (top 3)
# - person with most downvotes (top 3)

require 'json'
$base = "season-3"
data = JSON::load_file "#{$base}.json"
users = {}
User = Struct.new(:name, :img, :upvotes, :downvotes, :songs, :fans, :critics, :rank) {
    def initialize(name, img)
        super(
            name, img,
            0, 0,
            [],
            Hash.new(0), Hash.new(0),
            nil,
        )
    end
    
    def net_score
        upvotes + downvotes
    end
    
    def to_s
        [
            "User #{name}",
            "Net score #{net_score}",
            "Upvotes received #{upvotes}",
            "Downvotes received #{downvotes}",
        ].join "\n"
    end
    
    def to_json(*args)
        {
            name: name,
            img: img,
            rank: rank,
            net_score: net_score,
            upvotes: upvotes,
            downvotes: downvotes,
            critics: critics.sort_by { |key, score| score }.to_h,
            fans: fans.sort_by { |key, score| -score }.to_h,
            best_songs: songs.sort_by { |info| -info["votes"] },
            worst_songs: songs.sort_by { |info| info["votes"] },
        }.to_json(*args)
    end
}

data.each { |round_number, round|
    # p round_number
    round["info"].each { |song|
        user = song["submitter"]
        img = song["submitterImg"]
        user_info = (users[user] ||= User.new user, img)
        # puts "#{user} #{song["songName"]}"
        net_votes = 0
        song["responses"].each { |response|
            votes = response["votes"].to_i
            respondant = response["name"]
            net_votes += votes
            if votes < 0
                user_info.critics[respondant] += votes
                user_info.downvotes += votes
            elsif votes > 0
                user_info.fans[respondant] += votes
                user_info.upvotes += votes
            else
                # users who voted +0
            end
        }
        user_info.songs << {
            "name" => song["songName"],
            "votes" => net_votes,
            "round" => round_number,
            "img" => song["album"]["cover"]
        }
    }
}

def get_places(list, &cmp)
    cmp ||= -> a { a }
    indices = list
        .map
        .with_index
        .group_by { |a,i| cmp[a] }
        .sort_by { |key, val| key }
        .inject([]) { |build, (cmp_key, members)|
            last_rank = (build.last || [1]).first
            last_count = (build.last || [[]]).last.size
            [*build, [last_rank + last_count, members.map(&:last)]]
        }
    result = []
    indices.each { |rank, members| members.each { |idx| result[idx] = rank } }
    result
end

places = get_places(users.values) { |user| -user.net_score }

users.each.with_index { |(user, data), i|
    # make sure we have +0 for critics/fans
    users.each { |their_user, their_data|
        next if user == their_user
        their_data.critics[user] += 0
        their_data.fans[user] += 0
    }
    # give our user their rank
    data.rank = places[i]
}


# File.write "#{$base}-wrapped.json", users.to_json

user_list = users
    .sort_by { |username, user| user.name.downcase }
    .map { |username, user|
        "<li><a href=\"#!\" data-value=\"#{user.name}\"><img src=\"#{user.img}\" alt=\"#{user.name}\" class=\"left\">#{user.name}</a></li>"
    }
    .join "\n"

File.write "#{$base}-wrapped.html", <<EOT
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <!-- Add Materialize CSS and JavaScript CDN links -->
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/materialize/1.0.0/css/materialize.min.css">
  <script src="https://code.jquery.com/jquery-3.3.1.min.js"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/materialize/1.0.0/js/materialize.min.js"></script>
  <title>Season 3 Wrapped</title>
  <script>
    window.SeasonWrapped = #{users.to_json}
  </script>
  <link rel="stylesheet" href="style.css">
</head>
<body>

  <div class="container">
    <h3>Gaimes Music League Season 3 Wrapped</h3>

    <a class="dropdown-trigger btn" href="#" data-target="user-select"><i>Please select a user&hellip;</i></a>
    
    <ul id="user-select" class="dropdown-content">
      #{user_list}
    </ul>
    
    <div>
        <h4 id="selected-user"><i>Selected user will show here</i></h4>
        <div id="user-info"></div>
    </div>
  </div>

  <script src="wrapped.js"></script>

</body>
</html>

EOT

File: wrapped.js
const scoreToString = score =>
    (score >= 0 ? "+" : "") + score;

const getPlaces = (list, cmp=(a => a)) => {
    let grouped = new Map();
    list.forEach((element, index) => {
        let key = cmp(element);
        let val = grouped.get(key) ?? [];
        val.push({ element, index });
        grouped.set(key, val);
    });
    grouped = [...grouped];
    grouped.sort(([k1, v1], [k2, v2]) =>
        k1 < k2 ? 1 : k1 > k2 ? -1 : 0
    );
    let indices = grouped.reduce((build, [compare, members]) => {
        let lastRank = (build.at(-1) ?? [1])[0];
        let lastCount = (build.at(-1) ?? [[]]).at(-1).length;
        return [
            ...build,
            [
                lastRank + lastCount,
                members.map(member => member.index),
            ],
        ];
    }, []);
    let result = [];
    for(let [ rank, members ] of indices) {
        for(let idx of members) {
            result[idx] = rank;
        }
    }
    return result;
};

const compareCaseInsensitive = (a, b) =>
    a.localeCompare(b, undefined, { sensitivity: "base" })

const App = {
    setData(data) {
        this.data = data;
        this.processData();
        
        let userNames = Object.keys(this.data.members)
            .sort(compareCaseInsensitive);
        $("#user-select").empty();
        for(let name of userNames) {
            let a = $(`<a href="#!"></a>`);
            a.attr("data-value", name);
            let img = $(`<img class="left">`);
            img.attr("src", this.data.members[name]);
            img.attr("alt", name);
            a.append(img);
            a.append(name);
            let li = $("<li>");
            li.append(a);
            
            a.click(() => {
                App.loadData(a.attr("data-value"));
            });
            
            $("#user-select").append(li);
        }
        $("#app")[0].classList.toggle("hidden", !this.data);
    },
    
    processData() {
        this.stats = {
            users: new Map(),
        };
        for(let round of Object.values(this.data.rounds)) {
            for(let song of round.info) {
                let { submitter } = song;
                let userInfo = this.stats.users.get(submitter) ?? {
                    netScore: 0,
                    critics: {},
                    downvotes: 0,
                    fans: {},
                    upvotes: 0,
                    songs: [],
                };
                let netVotes = 0;
                for(let response of song.responses) {
                    let votes = parseInt(response.votes || "0", 10);
                    let respondant = response.name;
                    netVotes += votes;
                    if(votes < 0) {
                        userInfo.critics[respondant] ??= 0;
                        userInfo.critics[respondant] += votes;
                        userInfo.downvotes += votes;
                    }
                    else if(votes > 0) {
                        userInfo.fans[respondant] ??= 0;
                        userInfo.fans[respondant] += votes;
                        userInfo.upvotes += votes;
                    }
                    else {
                        // users who voted +0
                    }
                    userInfo.netScore += votes;
                }
                userInfo.songs.push({
                    name: song.songName,
                    votes: netVotes,
                    round: round.number,
                    img: song.album.cover
                });
                this.stats.users.set(submitter, userInfo);
            }
        }
        
        let places = getPlaces([...this.stats.users.values()], user => user.netScore);
        [...this.stats.users.entries()].forEach(([user, data], i) => {
            for(let [ theirUser, theirData ] of this.stats.users.entries()) {
                if(user !== theirUser) {
                    theirData.critics[user] ??= 0;
                    theirData.fans[user] ??= 0;
                }
            }
            data.rank = places[i];
        });
        for(let [ user, data] of this.stats.users.entries()) {
            data.bestSongs = [...data.songs].sort((s1, s2) => s2.votes - s1.votes);
            data.worstSongs = [...data.songs].sort((s1, s2) => s1.votes - s2.votes);
            data.critics = Object.entries(data.critics).sort((c1, c2) => c1[1] - c2[1]);
            data.fans = Object.entries(data.fans).sort((c1, c2) => c2[1] - c1[1]);
        }
    },
    
    userPreview(userName) {
        let userImg = this.data.members[userName];
        return $(`<a><img class="smol-icon" src="${userImg}" alt="${userName}">${userName}</a>`);
    },
    songPreview(songData) {
        return $(`<a><img class="smol-icon" src="${songData.img}" alt="${songData.name}">${songData.name}</a>`);
    },
    
    loadData(user) {
        console.log("Loading data for", user);
        let totalUsers = Object.entries(this.data.members).length;
        let userStats = this.stats.users.get(user);
        $("#selected-user")
            .empty()
            .append(this.userPreview(user));
        const critics = userStats.critics
            .map(([name, amount]) => $("<li>")
                .append(this.userPreview(name))
                .append(" " + scoreToString(amount))
            );
        const fans = userStats.fans
            .map(([name, amount]) => $("<li>")
                .append(this.userPreview(name))
                .append(" " + scoreToString(amount))
            );
        const bestSongs = userStats
            .bestSongs
            .slice(0, 5)
            .map(songData => $("<li>")
                .append(this.songPreview(songData))
                .append(" " + scoreToString(songData.votes) + ` (round ${songData.round})`)
            );
        const worstSongs = userStats
            .worstSongs
            .slice(0, 5)
            .map(songData => $("<li>")
                .append(this.songPreview(songData))
                .append(" " + scoreToString(songData.votes) + ` (round ${songData.round})`)
            );
        const newInfo = $(`
            <ul>
                <li style="flex: 0 1 100%;">Final score: ${scoreToString(userStats.netScore)} (${scoreToString(userStats.upvotes)}/${scoreToString(userStats.downvotes)}) &sdot; Placed ${userStats.rank}/${totalUsers}</li>
                <li style="flex: 2 0 40%;">Biggest critics: <ol class=critics></ol></li>
                <li style="flex: 2 0 40%;">Biggest fans: <ol class=fans></ol></li>
                <li style="flex: 2 0 40%;">Best songs: <ol class=bestsongs></ol></li>
                <li style="flex: 2 0 40%;">Worst songs: <ol class=worstsongs></ol></li>
            </ul>
        `);
        newInfo.find(".critics").append(critics);
        newInfo.find(".fans").append(fans);
        newInfo.find(".bestsongs").append(bestSongs);
        newInfo.find(".worstsongs").append(worstSongs);
        $("#user-info")
            .empty()
            .append(newInfo);
    },
    
    leagueAction(action) {
        if(action === "view-all-songs") {
            let allSongs = this.getAllSongs();
            $("#user-info").empty();
            let table = $("<table>");
            table.append("<tr><th>Song name</th><th>Song artist</th><th style=\"min-width: 20ch;\">Submitter(s)</th></tr>");
            let songs = [...allSongs.values()];
            songs.sort((s1, s2) =>
                compareCaseInsensitive(s1.submitters[0], s2.submitters[0])
                || compareCaseInsensitive(s1.artist, s2.artist)
                || compareCaseInsensitive(s1.songName, s2.songName)
            );
            for(let song of songs) {
                let songDisplay = $("<tr>");
                songDisplay.append($("<td>").append(this.songPreview({
                    img: song.album.cover,
                    name: song.songName,
                })));
                songDisplay.append($(`<td>${song.artist}</td>`));
                let submittersCell = $("<td>");
                song.submitters.forEach(submitter => {
                    submittersCell.append(this.userPreview(submitter));
                });
                songDisplay.append(submittersCell);
                table.append(songDisplay);
            }
            $("#user-info").append(table);
            // $("#user-info").text([...allSongs].map(([key, value]) => key));
        }
        else {
            alert(`Unknown action: ${action}`);
        }
    },
    
    getAllSongs() {
        let songMap = new Map();
        for(let round of Object.values(this.data.rounds)) {
            for(let submission of round.info) {
                let key = [ submission.artist, submission.songName, submission.album.name ];
                if(songMap.has(key)) {
                    let info = songMap.get(key);
                    info.submitters.push(submission.submitter);
                }
                else {
                    songMap.set(key, {
                        songName: submission.songName,
                        artist: submission.artist,
                        album: submission.album,
                        submitters: [ submission.submitter ],
                    });
                }
            }
        }
        return songMap;
    },
};


$(document).ready(function(){
    const jsonInput = $("#jsonFileInput");
    const onFileReady = () => {
        let file = jsonInput[0].files[0];
        if(file) {
            const reader = new FileReader();
            reader.onload = function(e) {
                const content = e.target.result;

                try {
                    const jsonData = JSON.parse(content);
                    App.setData(jsonData);
                }
                catch (error) {
                    console.error("Error parsing JSON:", error);
                }
            };
            reader.readAsText(file);
        }
    };
    jsonInput.on("change", onFileReady);
    onFileReady();
    
    $(".dropdown-trigger").dropdown();
    
    $("#league-info-select a").click((ev) => {
        App.leagueAction(ev.target.closest("a").dataset.value);
    });
});